# Лабораторная 4

In [1]:
import pandas as pd
data = pd.read_csv('./women-clothing-accessories.3-class.balanced.csv', sep='\t')

In [2]:
data.head()

review sentiment
0  качество плохое пошив ужасный (горловина напер...  negative
1  Товар отдали другому человеку, я не получила п...  negative
2  Ужасная синтетика! Тонкая, ничего общего с пре...  negative
3  товар не пришел, продавец продлил защиту без м...  negative
4      Кофточка голая синтетика, носить не возможно.  negative

In [3]:
data['sentiment'].value_counts()

negative    30000
neautral    30000
positive    30000
Name: sentiment, dtype: int64

In [4]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix

In [5]:
import re
import pymorphy2
lem = pymorphy2.MorphAnalyzer()
def cls(series: pd.Series) -> pd.Series:
    return series.apply(cl)
def cl(text: str) -> str:
    return clear(lemm(text))
def clear(text: str) -> str:
    return re.sub(r'[^\w\s]', ' ', text)
def lemm(text: str) -> str:
    return ' '.join(lem.parse(word)[0].normal_form for word in text.lower().split())

In [6]:
data['review']=cls(data['review'])

Разделение данных на обучающую и тестовую выборки

In [7]:
X = data['review']
Y = data['sentiment']
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

Преобразование текстов в числовой формат

In [8]:
vectorizer = TfidfVectorizer()
Xtrain = vectorizer.fit_transform(X_train)
Xtest = vectorizer.transform(X_test)

**Обучение модели логистической регрессии**

In [9]:
model = LogisticRegression(max_iter=1000)
model.fit(Xtrain, Y_train)

LogisticRegression(max_iter=1000)

Прогнозирование на тестовой выборке

In [10]:
Ypred = model.predict(Xtest)

Оценка качества модели

In [11]:
class_report = classification_report(Y_test, Ypred)

In [12]:
print('Classification Report:')
print(class_report)

Classification Report:
              precision    recall  f1-score   support

    neautral       0.63      0.64      0.64      6060
    negative       0.72      0.72      0.72      5942
    positive       0.87      0.85      0.86      5998

    accuracy                           0.74     18000
   macro avg       0.74      0.74      0.74     18000
weighted avg       0.74      0.74      0.74     18000



**Обучение модели случайного леса**

In [13]:
from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(n_estimators=50, max_depth=10, n_jobs=-1)
rf_model.fit(Xtrain, Y_train)

RandomForestClassifier(max_depth=10, n_estimators=50, n_jobs=-1)

Прогнозирование на тестовой выборке

In [14]:
ypred = rf_model.predict(Xtest)

Оценка качества модели 

In [15]:
class_report_rf = classification_report(Y_test, ypred)

In [16]:
print('Classification Report:')
print(class_report_rf)

Classification Report:
              precision    recall  f1-score   support

    neautral       0.54      0.58      0.56      6060
    negative       0.68      0.63      0.65      5942
    positive       0.74      0.74      0.74      5998

    accuracy                           0.65     18000
   macro avg       0.65      0.65      0.65     18000
weighted avg       0.65      0.65      0.65     18000



**Обучение модели Support Vector Machine** 

In [17]:
from sklearn.svm import LinearSVC
svm_model = LinearSVC()
svm_model.fit(Xtrain, Y_train)

LinearSVC()

Прогнозирование на тестовой выборке 

In [18]:
y_pred_svm = svm_model.predict(Xtest)

Оценка качества модели 

In [19]:
class_report_svm = classification_report(Y_test, y_pred_svm)

In [20]:
print('Support Vector Machine Classification Report:')
print(class_report_svm)

Support Vector Machine Classification Report:
              precision    recall  f1-score   support

    neautral       0.62      0.60      0.61      6060
    negative       0.69      0.71      0.70      5942
    positive       0.84      0.85      0.85      5998

    accuracy                           0.72     18000
   macro avg       0.72      0.72      0.72     18000
weighted avg       0.72      0.72      0.72     18000



### Итог

Точность:
1.  LogisticRegression - 0.74
2. RandomForestClassifier - 0.65
3. LinearSVC - 0.72

Наилучшая LogisticRegression

Deeppavlov не можем оценить(не хватает памяти)